src/core/async-work-cache.ts

```diff
@@ -31,11 +35,34 @@ export class AsyncWorkCache<K, V> {
    *
    * If not, compute it with the given function.
    *
-   * This method ensures that we will only try to compute the value for `key`
-   * once, no matter how often or with what timing getOrCompute is called, even
-   * recursively.
+   * This method ensures that, in the absence of cancellations, we will only try
+   * to compute the value for `key` once, no matter how often or with what
+   * timing getOrCompute is called, even recursively.
+   *
+   * If a compute() throws a Cancel that should not result in a Cancel for
+   * non-cancelled operations. So long as the given cancelToken is not
+   * cancelled, the caller will not receive a Cancel.
    */
-  async getOrCompute(key: K, compute: () => Promise<V>) {
+  async getOrCompute(
+      key: K, compute: () => Promise<V>, cancelToken?: CancelToken) {
+    cancelToken = cancelToken || neverCancels;
+    cancelToken.throwIfRequested();
+    while (true) {
+      try {
+        const result = await this._getOrCompute(key, compute);
+        cancelToken.throwIfRequested();
+        return result;
+      } catch (err) {
+        cancelToken.throwIfRequested();
```

Comment:
Should `cancelToken.throwIfRequested()` come last? In a case where there's both a cancel and a program exception, should the exception be thrown so we can see it?

Reply:
Hm. If you've cancelled an operation and then that operation fails, do you care? IMO, when you cancel an operation you no longer care about either success or failure.